scaffoldmaker/meshtypes/meshtype_3d_ellipticcylinder1.py

```diff
@@ -0,0 +1,476 @@
+
+    @staticmethod
+    def getDefaultOptions(parameterSetName='Default'):
+        return {'Number of elements around': 4, 'Number of elements up': 3, 'Number of elements through wall': 2,
+                'Major diameter': 1.2, 'Minor diameter': 0.7, 'Height': 2.0, 'Thickness ratio': 0.5,
+                'Use cross derivatives': False, 'Refine': False, 'Refine number of elements around': 1,
+                'Refine number of elements through wall': 1, 'Refine number of elements up': 1}
+
+    @staticmethod
+    def getOrderedOptionNames():
+        return ['Number of elements around', 'Number of elements up',
+                'Number of elements through wall', 'Major diameter', 'Minor diameter', 'Height', 'Thickness ratio',
+                'Use cross derivatives', 'Refine', 'Refine number of elements around', 'Refine number of elements up',
+                'Refine number of elements through wall']
+
+    @staticmethod
+    def checkOptions(options):
+        for key in ['Refine number of elements around', 'Refine number of elements up',
+                    'Refine number of elements through wall']:
+            if options[key] < 1:
+                options[key] = 1
+        if options['Number of elements up'] < 2:
+            options['Number of elements up'] = 2
+        if options['Number of elements around'] < 3:
+            options['Number of elements around'] = 3
+        if options['Number of elements through wall'] < 2:
+            options['Number of elements through wall'] = 2
+        if options['Major diameter'] < 0.0:
+            options['Diameter'] = 0.0
+        if options['Minor diameter'] < 0.0:
+            options['Diameter'] = 0.0
+        if options['Height'] < 0.0:
+            options['Height'] = 0.0
+        if options['Thickness ratio'] <= 0.0:
+            options['Thickness ratio'] = 0.01
+        elif options['Thickness ratio'] > 1.0:
+            options['Thickness ratio'] = 0.99
+
+    @classmethod
+    def generateBaseMesh(cls, region, options):
+        """
+        Generates the base tricubic Hermite mesh for this MeshType. See also generateMesh().
+        :param region: Zinc region to define model in. Must be empty.
+        :param options: Dict containing options. See getDefaultOptions().
+        :return: None
+        """
+        elementsCountAround = options['Number of elements around']
+        elementsCountUp = options['Number of elements up']
+        elementsCountThroughWall = options['Number of elements through wall']
+        useCrossDerivatives = options['Use cross derivatives']
+        axisA = options['Major diameter']
+        axisB = options['Minor diameter']
+        height = options['Height']
+        useCubicHermiteThroughWall = True
+
+        a = identity(4)
+
+        zero = [0.0, 0.0, 0.0]
+
+        """ Get nodes along the centre line """
+        cx = [zero, [0.0, 0.0, height]]
+        cd1 = [[0.0, 0.0, height / elementsCountUp], [0.0, 0.0, height / elementsCountUp]]
```

Comment:
Don't need to divide by elementsCountUp here for sampleCubicHermiteCurves to work as expected.

Reply:
Also, we've traditionally made xi1 go around meshes, xi2 go along meshes and xi3 go radially. In the other scaffolds this is followed, and s1, s2 and s3 do the same thing. This would mean your cd1 should be cd2, etc. Have a look at the 3d tube.

This may eventually change for long tube meshes such as the colon as we probably want nodes varying slowest along the tube as it's the most obvious partition for an efficient matrix.